src/coreclr/jit/emitarm64.cpp

```diff
@@ -6493,6 +6496,123 @@ void emitter::emitIns_R_R_R(
     appendToCurIG(id);
 }
 
+/*****************************************************************************
+ *
+ *  Add an instruction storing 2 registers into a memory (pointed by reg3) and the offset
+    (immediate).
+ */
+
+void emitter::emitIns_SS_R_R_R_I(instruction ins,
+                                 emitAttr    attr,
+                                 emitAttr    attr2,
+                                 regNumber   reg1,
+                                 regNumber   reg2,
+                                 regNumber   reg3,
+                                 ssize_t     imm,
+                                 int         varx1,
+                                 int         offs,
+                                 int         varx2)
+{
+    assert((ins == INS_stp) || (ins == INS_stnp));
+    emitAttr  size  = EA_SIZE(attr);
+    insFormat fmt   = IF_NONE;
+    unsigned  scale = 0;
+
+    // Is the target a vector register?
+    if (isVectorRegister(reg1))
+    {
+        assert(isValidVectorLSPDatasize(size));
+        assert(isVectorRegister(reg2));
+
+        scale = NaturalScale_helper(size);
+        assert((scale >= 2) && (scale <= 4));
+    }
+    else
+    {
+        assert(isValidGeneralDatasize(size));
+        assert(isGeneralRegisterOrZR(reg2));
+        scale = (size == EA_8BYTE) ? 3 : 2;
+    }
+
+    reg3 = encodingSPtoZR(reg3);
+
+    fmt          = IF_LS_3C;
+    ssize_t mask = (1 << scale) - 1; // the mask of low bits that must be zero to encode the immediate
+    if (imm == 0)
+    {
+        fmt = IF_LS_3B;
+    }
+    else
+    {
+        if ((imm & mask) == 0)
```

Comment:
What happens if this is false? In the `emitIns_R_R_S_S` case, we fall back to using the reserved register for reg3.

Reply:
seems like we never hit that path here. I will add `assert(false)` to test out.

Reply:
Also, for `ldr` (`emitIns_R_R_S_S()`) and `str` (`emitIns_S_S_R_R()`), I am passing a flag `useRegForImm` and asserting it with `assert(!useRegForImm)` whenever we do the replacement of ldr/str with ldp/stp to catch these scenarios. I want to see if there are any existing problems.